operator/builtin/output/googlecloud/request.go

```diff
@@ -0,0 +1,73 @@
+package googlecloud
+
+import (
+	"github.com/observiq/stanza/entry"
+	"go.uber.org/zap"
+
+	"google.golang.org/genproto/googleapis/api/monitoredres"
+	"google.golang.org/genproto/googleapis/logging/v2"
+)
+
+// RequestBuilder is an interface for creating requests to the google logging API
+type RequestBuilder interface {
+	Build(entries []*entry.Entry) []*logging.WriteLogEntriesRequest
+}
+
+// GoogleRequestBuilder builds google cloud logging requests
+type GoogleRequestBuilder struct {
+	MaxRequestSize int
+	ProjectID      string
+	EntryBuilder   EntryBuilder
+	*zap.SugaredLogger
+}
+
+// Build builds a series of write requests from stanza entries
+func (g *GoogleRequestBuilder) Build(entries []*entry.Entry) []*logging.WriteLogEntriesRequest {
+	currentSize := 0
+	protoEntries := []*logging.LogEntry{}
+	requests := []*logging.WriteLogEntriesRequest{}
+
+	for _, entry := range entries {
+		protoEntry, protoSize, err := g.EntryBuilder.Build(entry)
+		if err != nil {
+			g.Errorw("Failed to create protobuf entry. Dropping entry", zap.Any("error", err))
+			continue
+		}
+
+		if currentSize+protoSize > g.MaxRequestSize {
+			g.Debugw("Reached request size limit. Creating request.", "size", currentSize)
+			requests = append(requests, g.buildRequest(protoEntries))
+			protoEntries = []*logging.LogEntry{}
+			currentSize = 0
+		}
+
+		protoEntries = append(protoEntries, protoEntry)
+		currentSize += protoSize
```

Comment:
Is this a precise calculation? (i.e. Is the request size really the exact sum of its entry sizes?)

I'm guessing there is at least some overhead not being accounted for here. Should we perhaps use `proto.Size` to ensure we're still under the limit after building?

Even if marginally inaccurate, I think we would run into issues periodically when the `MaxRequestSize` is set to match Google's API limitation.

I'm honing in on this in part because this challenge was the reason for the previous (reactive) design. I fully support precomputation of size it's either A) perfectly accurate, or B) the maximum error can be quantified and determined to be acceptable.

Either way, perhaps we should have a test that exercises the size computation at the margin.

Reply:
I implemented a change that should look directly at the request size now. It's a combination of the old pattern and the new. Feel free to double check the splitting algorithm, in case I missed anything.